lib/internal/modules/esm/utils.js

```diff
@@ -0,0 +1,107 @@
+'use strict';
+const {
+  ArrayIsArray,
+  SafeSet,
+  SafeWeakMap,
+  ObjectFreeze,
+} = primordials;
+
+const {
+  ERR_VM_DYNAMIC_IMPORT_CALLBACK_MISSING,
+  ERR_INVALID_ARG_VALUE,
+} = require('internal/errors').codes;
+
+const { getOptionValue } = require('internal/options');
+
+const {
+  setImportModuleDynamicallyCallback,
+  setInitializeImportMetaObjectCallback
+} = internalBinding('module_wrap');
+const {
+  getModuleFromWrap,
+} = require('internal/vm/module');
+
+const callbackMap = new SafeWeakMap();
+function setCallbackForWrap(wrap, data) {
+  callbackMap.set(wrap, data);
+}
+
+let defaultConditions;
+function getDefaultConditions() {
+  if (defaultConditions === undefined) {
+    initializeDefaultConditions();
+  }
+  return defaultConditions;
+}
+
+let defaultConditionsSet;
+function getDefaultConditionsSet() {
+  if (defaultConditionsSet === undefined) {
+    initializeDefaultConditions();
+  }
+  return defaultConditionsSet;
+}
+
+function initializeDefaultConditions() {
+  const userConditions = getOptionValue('--conditions');
+  const noAddons = getOptionValue('--no-addons');
+  const addonConditions = noAddons ? [] : ['node-addons'];
+
+  defaultConditions = ObjectFreeze([
+    'node',
+    'import',
+    ...addonConditions,
+    ...userConditions,
+  ]);
+  defaultConditionsSet = new SafeSet(defaultConditions);
+}
+
+/**
+ * @param {string[]} [conditions]
+ * @returns {Set<string>}
+ */
+function getConditionsSet(conditions) {
+  if (conditions !== undefined && conditions !== getDefaultConditions()) {
+    if (!ArrayIsArray(conditions)) {
+      throw new ERR_INVALID_ARG_VALUE('conditions', conditions,
+                                      'expected an array');
+    }
+    return new SafeSet(conditions);
+  }
+  return getDefaultConditionsSet();
+}
```

Comment:
Conditional exports are a feature of packages, for either ESM or CommonJS: https://nodejs.org/api/packages.html#conditional-exports. Perhaps this whole file should just be moved up to `lib/internal/modules/utils.js`? Or create a separate `lib/internal/modules/conditions.js` for the conditions code, which makes up something like half of this file?

Reply:
For some reason CJS loader has its own copy of the conditions, so either this ends up in the `modules/helpers.js`, or stay in `esm/utils.js`. I prefer not to create another small module for this, adding small modules carelessly was part of the reason why we had been having a startup performance regression since v14 (snapshot does not make them free, the cost still adds up when there are too many of them, and the number of internal modules more than doubled from v14 to v18). I'd only add more modules when it helps to make the dependency less messy, not for creating a module's sake :(

Reply:
> For some reason CJS loader has its own copy of the conditions

That’s curious. Are they identical? Could we collapse into one? If they need to stay separate then you can ignore this comment.

Re splitting into files, Jacob and I find it nearly essential to keep the code organized. I understand that there’s a performance cost but that’s almost always the case when prioritizing human readability and maintainability; we accept that tradeoff every time we write JavaScript instead of C++, for instance. That said, if there’s a technical solution where maintainers like me can keep the organization of small files but the module snapshot could get larger files, that seems worth pursuing. I feel like simple concatenation would probably be insufficient, but surely we could use a bundler to merge together all the files of a top-level module (or _all_ our internal JavaScript files) to create fewer and bigger files to be loaded into the snapshot.

Reply:
> That’s curious. Are they identical? Could we collapse into one?

No, one of them starts with `['node', 'import']`, another is `['require', 'node']`.

> surely we could use a bundler to merge together all the files of a top-level module

I think that was rejected a long time ago because it makes debugging harder (we then have to ship source maps, it's already hard enough to get good stack trace in bug reports, which is also why we decided not to hide the internal frames in the stack trace). Bundling also makes the build harder, we'd probably have to maintain a self-contained bundler implemented without JS (ideally, using toolchains we already have, or there would be another problem of supporting those on certain platforms) , or there would be a chicken-and-egg problem on new platforms/architectures.

> that’s almost always the case when prioritizing human readability and maintainability

Yes, and in this case I don't think the a conditions module is worth it. Whether really small modules (<100 LOC) actually improves readability comes down to personal preference. I'd say we should go with merged utils by default, and only split out smaller modules when that actually makes a significant difference in readability, otherwise we are back to the slippery slope of gradually bloating the number of internal modules again.

Reply:
> Bundling also makes the build harder, we’d probably have to maintain a self-contained bundler implemented without JS

I was thinking we could potentially use `esbuild`, as it’s a Go binary that could live outside of our toolchain, kind like how Python already does. Assuming we could do whatever bundling we need to do via `esbuild`‘s CLI (rather than its Node JS API) then all we would need to do is just download the binary onto the box doing the building.

I also agree we can postpone this discussion for now. I understand the desire to maximize snapshot performance but I think we need to defer somewhat to the authors of various modules what kind of organization they need to maintain those APIs. I’d rather we investigate technical/build time solutions first before imposing “this file is too small” requirements on contributors.

Reply:
> but I think we need to defer somewhat to the authors of various modules what kind of organization they need to maintain those APIs.

I think so far I've only seen arguments of personal style preferences, instead of arguments of practical needs. What API needs demand splitting out conditions into a separate module, for example?  If splitting a module out eliminates a circular dependency, I'd agree it's a good idea, but if splitting a module out only suits someone's personal style preference better, I'd say that's not worth the cost. New contributors can always come along and have different personal style preference, but performance overhead stays.

> I’d rather we investigate technical/build time solutions first before imposing “this file is too small” requirements on contributors

I think the time spent on using bigger modules by default is much more trivial compared to investigation into such solutions. If someone thinks that investigating such solutions in exchange of unlimited small modules is a good investment of their time, they can do that. I think merging unnecessarily small modules in exchange of a faster startup is a better investment of my time, so that's what I'd do instead.

> One place where we might be able to find some low-hanging fruit in this regard is our internal dependencies. Stuff like Uncici could maybe have their own build pipeline (some probably already do, like cjs-module-lexer) and then if those dependencies provide a bundled single file, that should be optimized for snapshotting.

Yeah but this is already a bit off-topic (these modules are only used in certain use cases, but not essential to all the applications, so we probably don't need them in the built-in snapshot anyway).

Reply:
Folding this because

> If they need to stay separate then you can ignore this comment.

And they are different. Feel free to open a separate issue regarding bundling & creating small modules without hurting performance.